File: packages/api-types/src/ws/dto/linear.ts
/** Linear integration wire DTOs (APP-056). */

export type LinearRateLimitResourcePayload = {
  limit: number;
  used: number;
  remaining: number;
  /** Unix epoch milliseconds when the window resets. */
  reset: number;
};

export type LinearRateLimitPayload = {
  requests: LinearRateLimitResourcePayload | null;
  complexity: LinearRateLimitResourcePayload | null;
};

export type LinearStatusPayload = {
  connected: boolean;
  auth_method?: string | null;
  viewer_name?: string | null;
  viewer_email?: string | null;
  /** True when Atmos Hub session is required but missing (APP-056/057). */
  needs_hub_login?: boolean;
};

export type LinearLabelPayload = {
  name: string;
  color?: string | null;
};

export type LinearAssigneePayload = {
  name: string;
  avatar_url?: string | null;
};

export type LinearGithubRefPayload = {
  owner: string;
  repo: string;
  number: number;
  kind: string;
  url: string;
};

export type LinearIssuePayload = {
  id: string;
  identifier: string;
  title: string;
  url: string;
  description?: string | null;
  priority: number;
  state_name?: string | null;
  state_type?: string | null;
  project_name?: string | null;
  project_id?: string | null;
  team_id?: string | null;
  team_key?: string | null;
  labels: LinearLabelPayload[];
  assignee?: LinearAssigneePayload | null;
  github_refs: LinearGithubRefPayload[];
  created_at?: string | null;
  updated_at?: string | null;
};

export type LinearIssueListPagePayload = {
  issues: LinearIssuePayload[];
  has_next_page: boolean;
  end_cursor?: string | null;
};

export type LinearLinkPayload = {
  guid: string;
  workspace_guid: string;
  provider: string;
  external_id: string;
  identifier: string;
  title: string;
  url: string;
  snapshot_json?: string | null;
  linked_at: string;
};

File: apps/api/src/api/ws/message/linear.rs
use serde::{Deserialize, Serialize};

/// Hub auth for local → Hub integration calls (APP-056 / APP-057).
/// Prefer device credential (cross-origin safe); cookie is optional same-site fallback.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct HubSessionFields {
    /// Full `Cookie` header value for Hub origin (dev / same-site only).
    #[serde(default)]
    pub hub_cookie: Option<String>,
    /// Hub-minted device credential (plaintext). Sent as `Authorization: Bearer`.
    #[serde(default)]
    pub device_credential: Option<String>,
    /// Client-local Linear personal API key (never stored on Hub when used this way).
    #[serde(default)]
    pub linear_api_key: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinearConnectApiKeyRequest {
    pub api_key: String,
    #[serde(flatten)]
    pub hub: HubSessionFields,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinearOauthStartRequest {
    #[serde(default)]
    pub shell: Option<String>,
    #[serde(default)]
    pub web_origin: Option<String>,
    /// Public Linear OAuth client id (from web `NEXT_PUBLIC_LINEAR_OAUTH_CLIENT_ID`).
    #[serde(default)]
    pub client_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinearOauthFinishRequest {
    pub code: String,
    pub state: String,
    /// Same public client id used at start (PKCE token exchange).
    #[serde(default)]
    pub client_id: Option<String>,
    #[serde(flatten)]
    pub hub: HubSessionFields,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct LinearStatusRequest {
    #[serde(flatten)]
    pub hub: HubSessionFields,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct LinearRateLimitRequest {
    #[serde(flatten)]
    pub hub: HubSessionFields,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct LinearFilterOptionsRequest {
    #[serde(flatten)]
    pub hub: HubSessionFields,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct LinearDisconnectRequest {
    #[serde(flatten)]
    pub hub: HubSessionFields,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinearIssueListRequest {
    #[serde(default)]
    pub preset: Option<String>,
    #[serde(default)]
    pub team_id: Option<String>,
    #[serde(default)]
    pub project_id: Option<String>,
    #[serde(default)]
    pub query: Option<String>,
    #[serde(default)]
    pub first: Option<u32>,
    #[serde(default)]
    pub after: Option<String>,
    #[serde(flatten)]
    pub hub: HubSessionFields,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinearLinkIssueRequest {
    pub workspace_guid: String,
    pub issue: LinearIssueWire,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinearUnlinkIssueRequest {
    pub workspace_guid: String,
    pub external_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinearLinksForWorkspaceRequest {
    pub workspace_guid: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinearIssueWire {
    pub id: String,
    pub identifier: String,
    pub title: String,
    pub url: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub priority: i32,
    #[serde(default)]
    pub state_name: Option<String>,
    #[serde(default)]
    pub state_type: Option<String>,
    #[serde(default)]
    pub project_name: Option<String>,
    #[serde(default)]
    pub project_id: Option<String>,
    #[serde(default)]
    pub team_id: Option<String>,
    #[serde(default)]
    pub team_key: Option<String>,
    #[serde(default)]
    pub labels: Vec<LinearLabelWire>,
    #[serde(default)]
    pub assignee: Option<LinearAssigneeWire>,
    #[serde(default)]
    pub github_refs: Vec<LinearGithubRefWire>,
    #[serde(default)]
    pub created_at: Option<String>,
    #[serde(default)]
    pub updated_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinearLabelWire {
    pub name: String,
    #[serde(default)]
    pub color: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinearAssigneeWire {
    pub name: String,
    #[serde(default)]
    pub avatar_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LinearGithubRefWire {
    pub owner: String,
    pub repo: String,
    pub number: u64,
    pub kind: String,
    pub url: String,
}
